Allow passing custom Requestor class to Reddit.

elnuno opened the pull request:
This feature allows customizing the Reddit instance's Requestor. It can be useful for caching, logging and other enhancements. This, along with praw-dev/prawcore/pull/54 would make PRAW's networking more customizable. E.g., in `tests.integration`:

```python
    # Before

    reddit = Reddit(...)
    http = reddit._core._requestor._http
    recorder = Betamax(http)
```

```python
    # After, untested

    def req(*args, **kwargs):
        return Requestor(session_cls=betamaxxed, *args, **kwargs)

    def betamaxxed():
        return Betamax(requests.Session())

    reddit = Reddit(..., requestor_cls=req)

```

elnuno commented:
The suggestion on  praw-dev/prawcore#54 would help here, simplifying usage.

```python
    # After, untested

    def req(*args, **kwargs):
        return Requestor(session=Betamax(requests.Session()), *args, **kwargs)

    reddit = Reddit(..., requestor_cls=req)
```

I think it isn't feasible to pass a `Requestor` instance directly because `Reddit.config` is used to construct one (but it would probably be the cleanest design if doable). Simply passing a `session` to the constructor would solve one use  case, but using a customized Requestor is useful for adding behavior above the `requests.Session` level. One can memoize `Requestor.request`, dump request data to disk on `Requestor.close`, etc.

bboe commented:
Just curious what the status is on this PR now that prawcore has been updated.

elnuno commented:
I've used the approach of passing an already constructed Requestor, added a test (which I fear doesn't quite follow the clean style used) and a simple example. Glad to address any review requests and open to not merging if it's best :).

bboe commented on praw/reddit.py:
```diff
@@ -273,8 +276,11 @@ def _prepare_objector(self):
     def _prepare_prawcore(self):
-        requestor = Requestor(USER_AGENT_FORMAT.format(self.config.user_agent),
-                              self.config.oauth_url, self.config.reddit_url)
+        requestor = self.requestor
```
Could we use the `or` syntax like you did with prawcore? Assuming the parameter name is `custom_requestor`

    requestor = custom_requestor or Requestor(
        USER_AGENT_FORMAT...)

bboe replied:
I also wonder if there is a way we can pass in these default values so that USER_AGENT need not be explicitly set outside of the Reddit instance.

For example, rather than passing in a Requestor instance, or only a Requestor class, one can pass in a requestor class, in addition to keyword arguments to that class (e.g., to explicitly pass in a session).

That maybe would look like:

    reddit = praw.Reddit('sitename', requestor_class=JSONDebugRequestor, requestor_kwargs={'session': custom_http_session})

In fact, `requestor_class` could not be set (default to None) in the above example yet this provides a way to pass down the session.

elnuno replied:
Aha, this design eluded me! Changed it to passing a class and arguments.

bboe commented on tests/integration/test_reddit.py:
```diff
@@ -1,7 +1,13 @@
+
+from praw import Reddit
+from praw.const import USER_AGENT_FORMAT
+from prawcore import Requestor
```
To keep things grouped, can you please move the `prawcore` import to after `praw.models.reddit.base`?

bboe commented on tests/integration/test_reddit.py:
```diff
@@ -144,3 +150,24 @@ def test_top(self):
         assert len(submissions) == 100
+
+
+class TestRedditRequestor(IntegrationTest):
```
It looks like this could be a unit test as it doesn't actually result in issuing a request.

elnuno replied:
Done, moved it to tests/unit.

bboe commented on docs/examples/json_debbuging_reddit.py:
```diff
@@ -0,0 +1,35 @@
+import json
```
This examples directory is intended to contain completed scripts that are part of the documentation. Unless you want to include a tutorial on using this feature, I think a simple example in the code comments should suffice.

elnuno replied:
Added as a code comment. But if we figure out a nice PRAW debugging example (e.g., besides showing JSON, show some diagnostics or something like that), I'd be glad to add a tutorial and example in the future.

elnuno commented on praw/reddit.py:
```diff
@@ -84,6 +84,8 @@ def __init__(self, site_name=None, **config_settings):
             variable praw_site. If it is not found there, the DEFAULT site will
             be used.
 
+        :param requestor_cls: A class that will be used to handle HTTP requests.
```
Updated with the new params.

elnuno commented on praw/reddit.py:
```diff
@@ -130,6 +132,7 @@ def __init__(self, site_name=None, **config_settings):
                                   'must be set to None via a keyword arugment '
                                   'to the `Reddit` class constructor.')
 
+        self.requestor = requestor
```
Done, passing both requestor_class and requestor_kwargs, no use keeping them around.

elnuno commented:
Thank you for the encouragement, review and designs!

bboe commented on praw/reddit.py:
```diff
@@ -272,9 +277,32 @@ def _prepare_objector(self):
+        requestor_kwargs = requestor_kwargs or {}
+
+        # Allow overriding oauth_url and reddit_url as Requestor keywords
+        oauth_url = requestor_kwargs.pop('oauth_url', None)
```
Let's not make these values overwritable. If someone really want to change them, they can do so within the Requestor itself, but they're already changeable via the PRAW's configuration (praw.ini, env, or as arguments to `Reddit` itself.

bboe commented on praw/reddit.py:
```diff
@@ -272,9 +277,32 @@ def _prepare_objector(self):
+        reddit_url = requestor_kwargs.pop('reddit_url', None)
+        reddit_url = reddit_url or self.config.reddit_url
+
+        # Example usage:
```
I was thinking it would be great to make this part of the docstring for `Reddit` so that it shows up in the documentation. See examples for the `.. code:: python` for example here: https://github.com/praw-dev/praw/blob/d93d51c135432d6824793305b25c2cc64c0c0dc6/praw/models/reddit/submission.py#L51

I'm not sure why the ones in this file say `.. code-block`, maybe their aliases.

elnuno commented:
Reverted the overriding of oauth and reddit _url and moved the example into the docstring (but didn't built the docs).

I'll probably only be able to get back to this over the weekend, sorry. Thanks again for the review!

bboe commented on praw/reddit.py:
```diff
@@ -95,6 +100,25 @@ def __init__(self, site_name=None, **config_settings):
 
+        The ``requestor_class`` and ``requerstor_kwargs`` allow for
+        customization of the requestor :class`.Reddit` will use. This allows,
+        e.g., easily adding behavior to the requestor or wrapping it's
```
"its"

bboe commented on praw/reddit.py:
```diff
@@ -95,6 +100,25 @@ def __init__(self, site_name=None, **config_settings):
+        .. code-block:: python
+           import json, betamax, requests
+
+           class JSONDebugRequestor(Requestor):
```
Awesome. Thanks for taking the time to move this documentation here.

bboe commented on tests/unit/test_reddit.py:
```diff
@@ -128,3 +129,29 @@ def test_submission(self):
+        assert isinstance(reddit._core._requestor, CustomRequestor)
+        assert not isinstance(self.reddit._core._requestor, CustomRequestor)
+
+    def test_requestor_kwargs(self):
```
Great test!

bboe commented:
Merged as 572a82dc0870973ad6c23a12930fd5e31d003cb9. Thanks!

elnuno commented:
Woot! Dude, you rock! What a happy surprise on a Saturday night 👍
